FE may restart with failure when create MV with varchar partition column

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. CREATE MV 
```
CREATE MATERIALIZED VIEW mv_db.tb_d
PARTITION BY date_str
DISTRIBUTED BY hash(log_id)
PROPERTIES (
"force_external_table_query_rewrite" = "true"
) 
AS
select log_id, get_json_string(json_string,'$.col') as col_id,create_time, date_str
from hudi_catalog0.db_ods.tb 
where date_str= '2023-03-10' and hour_str = '10'
```
`date_str` type is varchar
2. Refresh MV
3. Restart FE

### Expected behavior (Required)
FE restart successfully

### Real behavior (Required)
```
java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) [starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
2023-03-10 14:58:04,244 WARN (stateChangeExecutor|81) [GlobalStateMgr.replayJournalInner():1913] catch exception when replaying 265966,
com.starrocks.journal.JournalException: fail to read journal entity key=265966, data=<DatabaseEntry offset="0" size="1368" data="0 12 0 0 0 0 0 0 44 91 0 0 0 0 0 0 87 29 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 0 0 0 0 0 0 87 32 0 0 0 9 112 50 48 50 50 49 48 50 48 0 0 0 6 78 79 82 77 65 76 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 0 0 0 0 0 0 87 30 0 0 0 6 78 79 82 77 65 76 0  ... 1268 bytes not shown "/>
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:242) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
Caused by: java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) ~[starrocks-fe.jar:?]
        ... 8 more
2023-03-10 14:58:04,245 ERROR (stateChangeExecutor|81) [GlobalStateMgr.transferToLeader():1036] failed to init journal after transfer to leader! will exit
com.starrocks.journal.JournalException: fail to read journal entity key=265966, data=<DatabaseEntry offset="0" size="1368" data="0 12 0 0 0 0 0 0 44 91 0 0 0 0 0 0 87 29 255 255 255 255 255 2568 bytes not shown "/>
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:242) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
Caused by: java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) ~[starrocks-fe.jar:?]
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

--- a/fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java
@@ -551,6 +551,7 @@ public class MaterializedViewAnalyzer {
                     throw new SemanticException("Materialized view partition column in partition exp " +
                             "must be base table partition column");
                 }
+                partitionColumns.forEach(partitionColumn1 -> checkPartitionColumnType(partitionColumn1));
             } else {
                 throw new SemanticException("Materialized view related base table partition type:" +
                         partitionInfo.getType().name() + "not supports");
@@ -558,14 +559,15 @@ public class MaterializedViewAnalyzer {
         }
 
         private void checkPartitionColumnWithBaseHMSTable(SlotRef slotRef, HiveMetaStoreTable table) {
-            List<String> partitionColumnNames = table.getPartitionColumnNames();
+            List<Column> partitionColumns = table.getPartitionColumns();
             if (table.isUnPartitioned()) {
                 throw new SemanticException("Materialized view partition column in partition exp " +
                         "must be base table partition column");
             } else {
                 boolean found = false;
-                for (String partitionColumn : partitionColumnNames) {
-                    if (partitionColumn.equalsIgnoreCase(slotRef.getColumnName())) {
+                for (Column partitionColumn : partitionColumns) {
+                    if (partitionColumn.getName().equalsIgnoreCase(slotRef.getColumnName())) {
+                        checkPartitionColumnType(partitionColumn);
                         found = true;
                         break;
                     }
@@ -586,8 +588,9 @@ public class MaterializedViewAnalyzer {
             } else {
                 boolean found = false;
                 for (PartitionField partitionField : partitionSpec.fields()) {
-                    String partitionColumn = partitionField.name();
-                    if (partitionColumn.equalsIgnoreCase(slotRef.getColumnName())) {
+                    String partitionColumnName = partitionField.name();
+                    if (partitionColumnName.equalsIgnoreCase(slotRef.getColumnName())) {
+                        checkPartitionColumnType(table.getColumn(partitionColumnName));
                         found = true;
                         break;
                     }
@@ -642,6 +645,14 @@ public class MaterializedViewAnalyzer {
                         break;
                     }
                 }
+            }
+        }
+
+        private void checkPartitionColumnType(Column partitionColumn) {
+            PrimitiveType type = partitionColumn.getPrimitiveType();
+            if (!type.isFixedPointType() && !type.isDateType()) {
+                throw new SemanticException("Materialized view partition exp column:"
+                        + partitionColumn.getName() + " with type "  + type + " not supported");
             }
         }
 

